// node-kakao/dist/api/auth-api-client.js
/*
 * Created on Thu Jan 28 2021
 *
 * Copyright (c) storycraft. Licensed under the MIT Licence.
 */
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
(function (factory) {
    if (typeof module === "object" && typeof module.exports === "object") {
        var v = factory(require, exports);
        if (v !== undefined) module.exports = v;
    }
    else if (typeof define === "function" && define.amd) {
        define(["require", "exports", "crypto", "./web-client", "../config", "../request", "./header-util", "./struct", "./xvc"], factory);
    }
})(function (require, exports) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    exports.AuthApiClient = exports.KnownAuthStatusCode = void 0;
    const crypto_1 = __importDefault(require("crypto"));
    const web_client_1 = require("./web-client");
    const config_1 = require("../config");
    const request_1 = require("../request");
    const header_util_1 = require("./header-util");
    const struct_1 = require("./struct");
    const xvc_1 = require("./xvc");
    /**
     * Status code for auth client results.
     */
    var KnownAuthStatusCode;
    (function (KnownAuthStatusCode) {
        KnownAuthStatusCode[KnownAuthStatusCode["INVALID_PHONE_NUMBER"] = 1] = "INVALID_PHONE_NUMBER";
        KnownAuthStatusCode[KnownAuthStatusCode["SUCCESS_WITH_ACCOUNT"] = 10] = "SUCCESS_WITH_ACCOUNT";
        KnownAuthStatusCode[KnownAuthStatusCode["SUCCESS_WITH_DEVICE_CHANGED"] = 11] = "SUCCESS_WITH_DEVICE_CHANGED";
        KnownAuthStatusCode[KnownAuthStatusCode["MISMATCH_PASSWORD"] = 12] = "MISMATCH_PASSWORD";
        KnownAuthStatusCode[KnownAuthStatusCode["EXCEED_LOGIN_LIMIT"] = 13] = "EXCEED_LOGIN_LIMIT";
        KnownAuthStatusCode[KnownAuthStatusCode["MISMATCH_PHONE_NUMBER"] = 14] = "MISMATCH_PHONE_NUMBER";
        KnownAuthStatusCode[KnownAuthStatusCode["EXCEED_PHONE_NUMBER_CHECK_LIMIT"] = 15] = "EXCEED_PHONE_NUMBER_CHECK_LIMIT";
        KnownAuthStatusCode[KnownAuthStatusCode["NOT_EXIST_ACCOUNT"] = 16] = "NOT_EXIST_ACCOUNT";
        KnownAuthStatusCode[KnownAuthStatusCode["NEED_CHECK_PHONE_NUMBER"] = 20] = "NEED_CHECK_PHONE_NUMBER";
        KnownAuthStatusCode[KnownAuthStatusCode["NEED_CHECK_QUIZ"] = 25] = "NEED_CHECK_QUIZ";
        KnownAuthStatusCode[KnownAuthStatusCode["DORMANT_ACCOUNT"] = 26] = "DORMANT_ACCOUNT";
        KnownAuthStatusCode[KnownAuthStatusCode["RESTRICTED_ACCOUNT"] = 27] = "RESTRICTED_ACCOUNT";
        KnownAuthStatusCode[KnownAuthStatusCode["LOGIN_FAILED"] = 30] = "LOGIN_FAILED";
        KnownAuthStatusCode[KnownAuthStatusCode["NOT_VERIFIED_EMAIL"] = 31] = "NOT_VERIFIED_EMAIL";
        KnownAuthStatusCode[KnownAuthStatusCode["MOBILE_UNREGISTERED"] = 32] = "MOBILE_UNREGISTERED";
        KnownAuthStatusCode[KnownAuthStatusCode["UNKNOWN_PHONE_NUMBER"] = 99] = "UNKNOWN_PHONE_NUMBER";
        KnownAuthStatusCode[KnownAuthStatusCode["SUCCESS_SAME_USER"] = 100] = "SUCCESS_SAME_USER";
        KnownAuthStatusCode[KnownAuthStatusCode["SUCCESS_SAME_USER_BY_MIGRATION"] = 101] = "SUCCESS_SAME_USER_BY_MIGRATION";
        KnownAuthStatusCode[KnownAuthStatusCode["TOO_MANY_REQUEST_A_DAY"] = -20] = "TOO_MANY_REQUEST_A_DAY";
        KnownAuthStatusCode[KnownAuthStatusCode["TOO_MANY_REQUEST_AT_A_TIME"] = -30] = "TOO_MANY_REQUEST_AT_A_TIME";
        KnownAuthStatusCode[KnownAuthStatusCode["MISMATCH_PASSCODE"] = -31] = "MISMATCH_PASSCODE";
        KnownAuthStatusCode[KnownAuthStatusCode["EXCEED_DAILY_REQUEST_LIMIT"] = -32] = "EXCEED_DAILY_REQUEST_LIMIT";
        KnownAuthStatusCode[KnownAuthStatusCode["EXCEED_DAILY_REQUEST_LIMIT_VOICECALL"] = -33] = "EXCEED_DAILY_REQUEST_LIMIT_VOICECALL";
        KnownAuthStatusCode[KnownAuthStatusCode["EXCEED_DAILY_REQUEST_LIMIT_WITHOUT_TOKEN"] = -34] = "EXCEED_DAILY_REQUEST_LIMIT_WITHOUT_TOKEN";
        KnownAuthStatusCode[KnownAuthStatusCode["DEVICE_NOT_REGISTERED"] = -100] = "DEVICE_NOT_REGISTERED";
        KnownAuthStatusCode[KnownAuthStatusCode["ANOTHER_LOGON"] = -101] = "ANOTHER_LOGON";
        KnownAuthStatusCode[KnownAuthStatusCode["DEVICE_REGISTER_FAILED"] = -102] = "DEVICE_REGISTER_FAILED";
        KnownAuthStatusCode[KnownAuthStatusCode["INVALID_DEVICE_REGISTER"] = -110] = "INVALID_DEVICE_REGISTER";
        KnownAuthStatusCode[KnownAuthStatusCode["INVALID_PASSCODE"] = -111] = "INVALID_PASSCODE";
        KnownAuthStatusCode[KnownAuthStatusCode["PASSCODE_REQUEST_FAILED"] = -112] = "PASSCODE_REQUEST_FAILED";
        KnownAuthStatusCode[KnownAuthStatusCode["NEED_TERMS_AGREE"] = -126] = "NEED_TERMS_AGREE";
        KnownAuthStatusCode[KnownAuthStatusCode["DENIED_DEVICE_MODEL"] = -132] = "DENIED_DEVICE_MODEL";
        KnownAuthStatusCode[KnownAuthStatusCode["RESET_STEP"] = -940] = "RESET_STEP";
        KnownAuthStatusCode[KnownAuthStatusCode["ACCESS_TOKEN_EXPIRED"] = -950] = "ACCESS_TOKEN_EXPIRED";
        KnownAuthStatusCode[KnownAuthStatusCode["NEED_PROTECTOR_AGREE"] = -991] = "NEED_PROTECTOR_AGREE";
        KnownAuthStatusCode[KnownAuthStatusCode["ACCOUNT_RESTRICTED"] = -997] = "ACCOUNT_RESTRICTED";
        KnownAuthStatusCode[KnownAuthStatusCode["INVALID_STAGE_ERROR"] = -998] = "INVALID_STAGE_ERROR";
        KnownAuthStatusCode[KnownAuthStatusCode["UPGRADE_REQUIRED"] = -999] = "UPGRADE_REQUIRED";
        KnownAuthStatusCode[KnownAuthStatusCode["VOICE_CALL_ONLY"] = -10002] = "VOICE_CALL_ONLY";
        KnownAuthStatusCode[KnownAuthStatusCode["ACCESSIBILITY_ARS_ONLY"] = -10003] = "ACCESSIBILITY_ARS_ONLY";
        KnownAuthStatusCode[KnownAuthStatusCode["MIGRATION_FAILURE"] = -100001] = "MIGRATION_FAILURE";
        KnownAuthStatusCode[KnownAuthStatusCode["INVALID_TOKEN"] = -100002] = "INVALID_TOKEN";
        KnownAuthStatusCode[KnownAuthStatusCode["UNDEFINED"] = -999999] = "UNDEFINED";
    })(KnownAuthStatusCode = exports.KnownAuthStatusCode || (exports.KnownAuthStatusCode = {}));
    function isXVCProvider(value) {
        return value !== null && typeof value === 'object' && typeof value.toFullXVCKey === 'function';
    }
    function toFiniteNumber(value) {
        const number = Number(value);
        return Number.isFinite(number) ? number : undefined;
    }
    function delay(ms, signal) {
        return new Promise((resolve, reject) => {
            if (signal === null || signal === void 0 ? void 0 : signal.aborted) {
                reject(signal.reason || new Error('Device registration aborted'));
                return;
            }
            const timer = setTimeout(() => {
                signal === null || signal === void 0 ? void 0 : signal.removeEventListener('abort', onAbort);
                resolve();
            }, ms);
            const onAbort = () => {
                clearTimeout(timer);
                reject(signal.reason || new Error('Device registration aborted'));
            };
            signal === null || signal === void 0 ? void 0 : signal.addEventListener('abort', onAbort, { once: true });
        });
    }
    /**
     * Android authentication client.
     *
     * The password login endpoint remains form encoded. New-device approval uses
     * the current passcodeLogin JSON flow: generate a challenge, approve it in the
     * official KakaoTalk app, poll registerDevice, then retry login.
     */
    class AuthApiClient {
        constructor(client, _name, _deviceUUID, advertisementIdOrConfig, configOrProvider, xvcProvider) {
            this._name = _name;
            this._deviceUUID = _deviceUUID;
            if (typeof advertisementIdOrConfig === 'string') {
                this._advertisementId = advertisementIdOrConfig;
                if (isXVCProvider(configOrProvider) && xvcProvider === undefined) {
                    this.config = Object.assign({ ...config_1.DefaultConfiguration }, {});
                    this.xvcProvider = configOrProvider;
                }
                else {
                    this.config = Object.assign({ ...config_1.DefaultConfiguration }, configOrProvider || {});
                    this.xvcProvider = xvcProvider || xvc_1.AndroidSubXVCProvider;
                }
            }
            else {
                // Backward compatibility with the published declaration and existing callers:
                // new AuthApiClient(client, name, uuid, config, xvcProvider)
                this._advertisementId = crypto_1.default.randomUUID();
                this.config = Object.assign({ ...config_1.DefaultConfiguration }, advertisementIdOrConfig || {});
                this.xvcProvider = isXVCProvider(configOrProvider)
                    ? configOrProvider
                    : (xvcProvider || xvc_1.AndroidSubXVCProvider);
            }
            this._client = new web_client_1.DataWebRequest(client);
            this._activePasscodeChallenge = undefined;
        }
        get name() {
            return this._name;
        }
        get deviceUUID() {
            return this._deviceUUID;
        }
        get advertisementId() {
            return this._advertisementId;
        }
        async createAuthHeader(form, contentType) {
            const header = {};
            (0, header_util_1.fillBaseHeader)(header, this.config);
            (0, header_util_1.fillAHeader)(header, this.config);
            const userAgent = (0, header_util_1.getUserAgent)(this.config);
            header['User-Agent'] = userAgent;
            header['Adid'] = this._advertisementId;
            header['X-Vc'] = await this.calculateXVCKey(this.deviceUUID, userAgent, form.email);
            if (contentType)
                header['Content-Type'] = contentType;
            return header;
        }
        fillAuthForm(form) {
            const result = { ...form };
            result['one_store'] = result['one_store'] === undefined ? false : result['one_store'];
            result['permanent'] = result['permanent'] === undefined ? true : result['permanent'];
            result['device_uuid'] = this._deviceUUID;
            result['device_name'] = this._name;
            if (this.config.deviceModel)
                result['model_name'] = this.config.deviceModel;
            result['forced'] = result['forced'] === undefined ? false : result['forced'];
            return result;
        }
        /** Login using an email/password credential. */
        async login(form) {
            const res = await this._client.requestData('POST', this.getApiPath('login.json'), this.fillAuthForm(form), await this.createAuthHeader(form));
            const status = toFiniteNumber(res.status) ?? KnownAuthStatusCode.UNDEFINED;
            if (status !== request_1.KnownDataStatusCode.SUCCESS)
                return { status, success: false };
            return {
                status,
                success: true,
                result: (0, struct_1.structToLoginData)(res, this._deviceUUID),
            };
        }
        /**
         * Generate the passcode shown by the sub-device.
         * The user must approve this code in the official KakaoTalk app.
         */
        async generatePasscode(form, options = {}) {
            const body = {
                email: form.email,
                password: form.password,
                permanent: options.permanent === undefined ? true : options.permanent,
                device: {
                    name: this._name,
                    uuid: this._deviceUUID,
                    model: this.config.deviceModel || this._name,
                    osVersion: options.deviceOsApiLevel || '35',
                    isOneStore: false,
                },
            };
            const res = await this._client.requestData('POST', this.getApiPath('passcodeLogin/generate'), body, await this.createAuthHeader(form, 'application/json; charset=utf-8'));
            const status = toFiniteNumber(res.status) ?? KnownAuthStatusCode.UNDEFINED;
            const passcode = typeof res.passcode === 'string'
                ? res.passcode
                : (typeof res.passcode === 'number' ? String(res.passcode) : undefined);
            if (status !== request_1.KnownDataStatusCode.SUCCESS || !passcode)
                return { status: status === request_1.KnownDataStatusCode.SUCCESS ? KnownAuthStatusCode.PASSCODE_REQUEST_FAILED : status, success: false };
            const challenge = {
                passcode,
                remainingSeconds: toFiniteNumber(res.remainingSeconds),
            };
            this._activePasscodeChallenge = challenge;
            return { status, success: true, result: challenge };
        }
        /**
         * Backward-compatible passcode request wrapper.
         *
         * Legacy callers expect the code to arrive externally and prompt on stdin.
         * The current Android flow returns the code to this client instead. When no
         * callback is supplied, print the challenge so those callers can still
         * complete official-app approval without changing their application code.
         */
        async requestPasscode(form, options = {}) {
            const result = await this.generatePasscode(form, options);
            if (!result.success)
                return result;
            if (typeof options.onPasscodeRequired === 'function') {
                await options.onPasscodeRequired(result.result);
            }
            else if (options.silent !== true && typeof console !== 'undefined' && typeof console.info === 'function') {
                console.info(`[node-kakao] 새 기기 인증 코드: ${result.result.passcode}`);
                console.info('[node-kakao] 공식 카카오톡에서 이 코드를 승인한 뒤, 기존 입력 프롬프트에서 Enter를 누르세요.');
            }
            return result;
        }
        /** Perform one registration-status poll. */
        async registerPasscodeDevice(form) {
            const body = {
                email: form.email,
                password: form.password,
                device: { uuid: this._deviceUUID },
            };
            const res = await this._client.requestData('POST', this.getApiPath('passcodeLogin/registerDevice'), body, await this.createAuthHeader(form, 'application/json; charset=utf-8'));
            const status = toFiniteNumber(res.status) ?? KnownAuthStatusCode.UNDEFINED;
            return {
                status,
                success: status === request_1.KnownDataStatusCode.SUCCESS,
                nextRequestIntervalInSeconds: toFiniteNumber(res.nextRequestIntervalInSeconds),
                remainingSeconds: toFiniteNumber(res.remainingSeconds),
            };
        }
        /** Cancel an outstanding passcode challenge. */
        async cancelDeviceRegistration(form) {
            const body = {
                email: form.email,
                password: form.password,
                device: { uuid: this._deviceUUID },
            };
            const res = await this._client.requestData('POST', this.getApiPath('passcodeLogin/cancel'), body, await this.createAuthHeader(form, 'application/json; charset=utf-8'));
            const status = toFiniteNumber(res.status) ?? KnownAuthStatusCode.UNDEFINED;
            return { status, success: status === request_1.KnownDataStatusCode.SUCCESS };
        }
        /**
         * Wait for main-device approval and register this device.
         *
         * The passcode and permanent parameters are retained for source
         * compatibility. The current server contract does not send the passcode
         * back to registerDevice; approval is observed by polling.
         */
        async registerDevice(form, _passcode, _permanent = true, options = {}) {
            let remainingSeconds = options.timeoutSeconds
                ?? (this._activePasscodeChallenge && this._activePasscodeChallenge.remainingSeconds)
                ?? 120;
            while (remainingSeconds > 0) {
                if (options.signal && options.signal.aborted)
                    throw options.signal.reason || new Error('Device registration aborted');
                const attempt = await this.registerPasscodeDevice(form);
                if (attempt.success) {
                    this._activePasscodeChallenge = undefined;
                    return { status: attempt.status, success: true };
                }
                if (attempt.status !== KnownAuthStatusCode.DEVICE_NOT_REGISTERED)
                    return { status: attempt.status, success: false };
                const intervalSeconds = attempt.nextRequestIntervalInSeconds === undefined
                    ? 3
                    : Math.max(0, attempt.nextRequestIntervalInSeconds);
                const serverRemaining = attempt.remainingSeconds;
                if (serverRemaining !== undefined)
                    remainingSeconds = Math.min(remainingSeconds, serverRemaining);
                remainingSeconds -= intervalSeconds;
                if (remainingSeconds <= 0)
                    break;
                await delay(intervalSeconds * 1000, options.signal);
            }
            try {
                await this.cancelDeviceRegistration(form);
            }
            catch (_) {
                // Cancellation is best-effort; retain the registration timeout status.
            }
            return { status: KnownAuthStatusCode.DEVICE_NOT_REGISTERED, success: false };
        }
        async calculateXVCKey(deviceUUID, userAgent, email) {
            return (await this.xvcProvider.toFullXVCKey(deviceUUID, userAgent, email)).substring(0, 16);
        }
        getApiPath(api) {
            return `${this.config.agent}/account/${api}`;
        }
        /**
         * Create an AuthApiClient.
         *
         * Supports both signatures:
         * - create(name, uuid, config?, xvcProvider?)
         * - create(name, uuid, advertisementId, config?, xvcProvider?)
         */
        static async create(name, deviceUUID, advertisementIdOrConfig = {}, configOrProvider, xvcProvider) {
            let advertisementId;
            let config;
            let provider;
            if (typeof advertisementIdOrConfig === 'string') {
                advertisementId = advertisementIdOrConfig;
                if (isXVCProvider(configOrProvider) && xvcProvider === undefined) {
                    config = {};
                    provider = configOrProvider;
                }
                else {
                    config = configOrProvider || {};
                    provider = xvcProvider;
                }
            }
            else {
                advertisementId = crypto_1.default.randomUUID();
                config = advertisementIdOrConfig || {};
                provider = isXVCProvider(configOrProvider) ? configOrProvider : xvcProvider;
            }
            return new AuthApiClient(await (0, web_client_1.createWebClient)('https', 'katalk.kakao.com'), name, deviceUUID, advertisementId, config, provider || xvc_1.AndroidSubXVCProvider);
        }
    }
    exports.AuthApiClient = AuthApiClient;
});
